src/cli/index.js

```diff
@@ -234,68 +235,129 @@ export function main({
     });
   };
 
-  //
   const runEventuallyWithNetwork = (mutexPort: ?string): Promise<void> => {
-    return new Promise(resolve => {
+    return new Promise((resolve, reject) => {
       const connectionOptions = {
         port: +mutexPort || constants.SINGLE_INSTANCE_PORT,
       };
 
-      const clients = new Set();
-      const server = net.createServer();
+      function startServer() {
+        const clients = new Set();
+        const server = http.createServer(manager);
 
-      server.on('error', () => {
-        // another Yarn instance exists, let's connect to it to know when it dies.
-        reporter.warn(reporter.lang('waitingInstance'));
-        const socket = net.createConnection(connectionOptions);
+        // The server must not prevent us from exiting
+        server.unref();
 
-        socket
-          .on('connect', () => {
-            // Allow the program to exit if this is the only active server in the event system.
-            socket.unref();
-          })
-          .on('close', (hadError?: boolean) => {
-            // the `close` event gets always called after the `error` event
-            if (!hadError) {
-              process.nextTick(() => {
-                resolve(runEventuallyWithNetwork(mutexPort));
-              });
-            }
-          })
-          .on('error', () => {
-            // No server to listen to ? Let's retry to become the next server then.
-            process.nextTick(() => {
-              resolve(runEventuallyWithNetwork(mutexPort));
-            });
+        // No socket must timeout
+        server.timeout = 0;
+
+        // If we fail to setup the server, we ask the existing one for its name
+        server.on('error', () => {
+          reportServerName();
+        });
+
+        // If we succeed, keep track of all the connected sockets to delete them later
+        server.on('connection', socket => {
+          clients.add(socket);
+          socket.on('close', () => {
+            clients.delete(socket);
           });
-      });
+        });
+
+        server.listen(connectionOptions, () => {
+          // Don't forget to kill the sockets if we're being killed via signals
+          onDeath(killSockets);
+
+          // Also kill the sockets if we finish, whether it's a success or a failure
+          run().then(
+            res => {
+              killSockets();
+              resolve(res);
+            },
+            err => {
+              killSockets();
+              reject(err);
+            },
+          );
+        });
+
+        function manager(request, response) {
+          response.writeHead(200);
+          response.write(config.cwd);
+          response.end();
```

Comment:
You can combine this with the line above: `response.end(config.cwd)` - https://nodejs.org/api/http.html#http_response_end_data_encoding_callback

That said, I'm a bit concerned about this part since `response.end` would close the connection. Why don't we just do `response.write()`, omit the `response.end()` call? That way we wouldn't need `reportServerName` and `waitForNetwork` together. We can merge their logic into one.

Reply:
Doing this would require us to implement our own protocol on top of HTTP to detect that we received the full server message and that we can print it, plus 'hacking' the HTTP protocol to keep the request alive, which is not usual. As simple as it can be, I strongly prefer keeping the logic separated: one function to get information, one function to get a keepalive.

Reply:
Ah, that's a very good reason: we don't have any means to detect the actual flushing. Looks like the ideal solution here would have been websockets, if they came built-in to node.js

Anyway, I think a short comment above the `response.end()` call saying something like the following would be great:```
// Call `.end()` to ensure the response is flushed.
```

That said I read the docs for `response.write()` and it reads like node.js always flushes each write: https://nodejs.org/api/http.html#http_response_write_chunk_encoding_callback

I'm not unhappy with the current implementation, I just find it a bit hairy and hacky since our second connection is a raw TCP connection and relies on the client technically DoS'ing the mutex by not responding with an ACK.